File: src/main/java/tk/skulk/plugin/core/extensions/updatechecker/UpdateCheckerExtension.java
package tk.skulk.plugin.core.extensions.updatechecker;

import okhttp3.Request;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import tk.skulk.plugin.core.Plugin;
import tk.skulk.plugin.core.extensions.updatechecker.commands.UpdateCommand;
import tk.skulk.plugin.core.extensions.updatechecker.listeners.OpJoinListener;
import tk.skulk.plugin.singletons.Singletons;
import tk.skulk.plugin.wrapper.BaseExtension;

public final class UpdateCheckerExtension extends BaseExtension {
    private static final @NotNull Request REQUEST = new Request.Builder()
            .url("https://api.github.com/repos/Skulk-Network/Plugin/releases/latest")
            .get()
            .build();

    @SuppressWarnings("ConstantConditions")
    public @Nullable String getLatestVersion() {
        try (final var response = Singletons.OKHTTP.newCall(UpdateCheckerExtension.REQUEST).execute()) {
            return Singletons.GSON.fromJson(response.body().string(), UpdateCheckerExtension.ReleaseTagNameResponse.class).tag_name;
        } catch (final Exception error) {
            return null;
        }
    }

    public UpdateCheckerExtension(final @NotNull Plugin plugin) {
        super(plugin);
    }

    @Override protected void initCommands() {
        new UpdateCommand(this);
    }

    @Override protected void initListeners() {
        new OpJoinListener(this);
    }

    private static final class ReleaseTagNameResponse {
        private String tag_name;
    }
}

File: src/main/kotlin/tk/skulk/plugin/core/extensions/updateChecker/UpdateCheckerExtension.kt
package tk.skulk.plugin.core.extensions.updateChecker

import com.google.gson.annotations.SerializedName
import okhttp3.Request
import tk.skulk.plugin.core.SNPlugin
import tk.skulk.plugin.core.extensions.updateChecker.commands.UpdateCommand
import tk.skulk.plugin.core.extensions.updateChecker.listeners.OpJoinListener
import tk.skulk.plugin.core.gson
import tk.skulk.plugin.core.http
import tk.skulk.plugin.framework.Extension

val request = Request
    .Builder()
    .url("https://api.github.com/repos/Skulk-Network/Plugin/releases/latest")
    .get()
    .build()

class ReleaseTagNameResponse(
    @SerializedName("tag_name")
    val tagName: String,
)

class UpdateCheckerExtension(plugin: SNPlugin) : Extension<SNPlugin>(plugin) {
    override val commands = listOf {
        UpdateCommand(this)
    }

    override val listeners = listOf {
        OpJoinListener(this)
    }

    val latestVersion: String
        get() = http.newCall(request).execute().use { response ->
            gson.fromJson(
                response.body!!.string(), ReleaseTagNameResponse::class.java
            )!!.tagName
        }
}
